### PR TITLE
Draft createComponent extension point

### PR DESCRIPTION
**Pull request by owaiskazi19**
Signed-off-by: Owais Kazi <owaiskazi19@gmail.com>

### Description
_Describe what this change achieves._

### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._ 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by joshpalis — DISMISSED** [review]

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 1]
```diff
@@ -90,12 +95,47 @@ public static ExtensionSettings getExtensionSettings() throws IOException {
...
+                clusterSettingResponseHandler
+            );
+            LocalNodeResponseHandler localNodeResponseHandler = new LocalNodeResponseHandler();
+            transportService.sendRequest(
+                opensearchNode,
+                ExtensionsOrchestrator.REQUEST_EXTENSION_LOCAL_NODE,
+                new ExtensionRequest(ExtensionsOrchestrator.RequestType.REQUEST_EXTENSION_LOCAL_NODE),
+                localNodeResponseHandler
+            );
+            inProgressLatch.await(1, TimeUnit.SECONDS);
```
Comment (peternied) [inline src/main/java/org/opensearch/sdk/ExtensionsRunner.java:132]: Looks like this CountDownLatch is only used to wait 1 second, and it isn't counted down.  The latch should be updated when the responses come back, I would suggest the count is 3 and its expected each of the `sendRequest` methods are responsible for lowering the count by one.

I would recommend reauthoring this code to leverage the collections objects from future, `allOf(....)`.  If this doesn't work it might be worthwhile to invest in writing helpers to make declaring code following this pattern smaller and more error resistance since I think this example will be replicated in many places.
```java
CompletableFuture<?> clusterStateRequest =  transportServiceProvider.startRequest(....);
CompletableFuture<?> clusterSettingsRequest =  transportServiceProvider.startRequest(....);
CompletableFuture<?> localNodeRequest =  transportServiceProvider.startRequest(....);
try {
   // Result isn't used, timeout or inner exceptions are handled in catch block
   CompletableFuture.allOf(clusterStateRequest, clusterSettingsRequest, localNodeRequest).get(1, TimeUnit.SECONDS).
} catch (Exception e) {
   logger.error(e);
}
```

File: src/main/java/org/opensearch/sdk/LocalNodeResponseHandler.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public void handleResponse(LocalNodeResponse response) {
+        logger.info("received {}", response);
+        inProgressLatch.countDown();
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        logger.debug(new ParameterizedMessage("LocalNodeRequest failed"), exp);
+        // inProgressLatch.countDown();
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/LocalNodeResponseHandler.java:28]: Why is this commented out?

File: src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public void handleResponse(ClusterSettingsResponse response) {
+        logger.info("received {}", response);
+        inProgressLatch.countDown();
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        logger.debug(new ParameterizedMessage("ClusterSettingRequest failed"), exp);
+        // inProgressLatch.countDown();
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java:28]: Why is this commented out?

File: src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public void handleResponse(ExtensionClusterStateResponse response) {
+        logger.info("received {}", response);
+        inProgressLatch.countDown();
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        logger.debug(new ParameterizedMessage("ClusterStateRequest failed"), exp);
+        // inProgressLatch.countDown();
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java:28]: Why is this commented out?

File: src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public void handleResponse(ExtensionClusterStateResponse response) {
+        logger.info("received {}", response);
+        inProgressLatch.countDown();
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        logger.debug(new ParameterizedMessage("ClusterStateRequest failed"), exp);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java:27]: Seems odd to have an exception message at a lower logging level (debug) than the expected response (info).

File: src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.cluster.ExtensionClusterStateResponse;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponseHandler;
+
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+
+public class ClusterStateResponseHandler implements TransportResponseHandler<ExtensionClusterStateResponse> {
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java:15]: Should the class name be `ExtensionClusterStateResponseHandler`?

File: src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public void handleResponse(ClusterSettingsResponse response) {
+        logger.info("received {}", response);
+        inProgressLatch.countDown();
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        logger.debug(new ParameterizedMessage("ClusterSettingRequest failed"), exp);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java:27]: Why wrap the String in a ParameterizedMessage instead of [passing directly](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/Logger.html#debug-java.lang.String-java.lang.Object...-)?

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 8]
```diff
@@ -90,12 +95,47 @@ public static ExtensionSettings getExtensionSettings() throws IOException {
     PluginResponse handlePluginsRequest(PluginRequest pluginRequest) {
         logger.info("Handling Plugins Request");
         PluginResponse pluginResponse = new PluginResponse("RealExtension");
+        opensearchNode = pluginRequest.getSourceNode();
         return pluginResponse;
     }
 
-    IndicesModuleResponse handleIndicesModuleRequest(IndicesModuleRequest indicesModuleRequest) {
+    IndicesModuleResponse handleIndicesModuleRequest(IndicesModuleRequest indicesModuleRequest, TransportService transportService) {
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: Created a setter for TransportService and it's available when needed.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 1 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: Removed CountDown Latch as it was not needed here.

File: src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    private static final Logger logger = LogManager.getLogger(ClusterStateResponseHandler.class);
+    final CountDownLatch inProgressLatch = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java]: Created this handler only during the bootstrap just to make sure ClusterState is accessible from OpenSearch. Since SDK will work as a boiler plate and later based on whichever plugin integrates SDK will create their own handler for on their use case.

File: src/main/java/org/opensearch/sdk/ExtensionClusterStateResponseHandler.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public void handleResponse(ExtensionClusterStateResponse response) {
+        logger.info("received {}", response);
+        inProgressLatch.countDown();
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        logger.info("ClusterStateRequest failed", exp);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ExtensionClusterStateResponseHandler.java:26]: Is this "ExtensionClusterStateRequest"?

**Review by dbwiddis — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponseHandler;
+
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+
+public class ClusterSettingResponseHandler implements TransportResponseHandler<ClusterSettingsResponse> {
+    private static final Logger logger = LogManager.getLogger(ClusterSettingResponseHandler.class);
+    final CountDownLatch inProgressLatch = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (peternied) [reply src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java:16]: What is the purpose of this object, it doesn't seem to be impacted flow of control?  Here is a stack overflow answer about [CountDownLatch's usage](https://stackoverflow.com/a/17827339/533057).

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 1 above)
Comment (peternied) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: I'm confused, don't this method need to wait until the transport requests have finished?  If we remove a mechanism to wait we aren't able to handle failures where the extension wasn't able to honor one of the requests and send back error information.

If these requests are executed synchronously, I suppose the current code make sense, but it slows down the response considerable if any of these actions can be done in parallel.  Since none of the responses are used as the input to later requests this seems to be the case.

Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: The latch was added to make the calls asyn before but since now each request has their own handler they have their own latch and can be handled separately. @saratvemulapalli correct me I'm wrong here

File: src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ClusterSettingResponseHandler.java:16]: https://github.com/opensearch-project/opensearch-sdk/pull/8#discussion_r899630182

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 1 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: Removed the latch here as it was a hacky way of making Transport call and from the handler as well.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 12]
```diff
@@ -56,10 +57,9 @@
 import static org.opensearch.common.UUIDs.randomBase64UUID;
 
 public class ExtensionsRunner {
-
-    public static final String REQUEST_EXTENSION_ACTION_NAME = "internal:discovery/extensions";
-
     private static ExtensionSettings extensionSettings = null;
+    private DiscoveryNode opensearchNode = null;
+    TransportService transportService = null;
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: Good call! Made transportService private and created a getter for the same.

File: src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ClusterStateResponseHandler.java]: SDK will consist of all the extension points mentioned [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/Plugin.java). Based on the requirement of the plugin they will access the extension points when integrating with SDK and making their plugin to work as an extension. These extension points will provide a hook to connect the extension with OpenSearch code through Transport Request/Response.

ClusterState, ClusterSettings, LocalNode will be required by createComponent extension point.
More details on the design is [here](https://github.com/opensearch-project/OpenSearch/issues/2447).

**Review by dbwiddis — CHANGES_REQUESTED** [review]
TLDR: I'm not sure what the right answer is with TransportService, but I don't think what we have is it.

That pesky TransportService instance variable needs a bit more careful thought and handling here.  You're both setting it and returning it with the `create()` method, and elsewhere using both effects (return and instance) interchangeably.

If we need to keep the same API (with a return value from `create*()` and an argument to `start*()`) then we should find a more consistent way to work within those constraints (perhaps not have the instance variable and require the user to create and then pass it).

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 13]
```diff
@@ -56,52 +57,79 @@
 import static org.opensearch.common.UUIDs.randomBase64UUID;
 
 public class ExtensionsRunner {
+    private ExtensionSettings extensionSettings = getExtensionSettings();
+    private DiscoveryNode opensearchNode;
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:61]: Created setter/getter to access opensearchNode.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 14]
```diff
@@ -56,52 +57,79 @@
...
+    private TransportService transportService;
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: Removed instance variable `transportService` and passed as an arguments

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 15]
```diff
@@ -188,7 +223,7 @@ public void startTransportService(TransportService transportService) {
             false,
             false,
             IndicesModuleRequest::new,
-            ((request, channel, task) -> channel.sendResponse(handleIndicesModuleRequest(request)))
+            ((request, channel, task) -> channel.sendResponse(handleIndicesModuleRequest(request, transportService)))
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/ExtensionsRunner.java:226]: We'll need to pass transportService here for the API and also for the tests. Since instance variable is removed

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 16]
```diff
@@ -56,52 +57,79 @@
...
-    private static final Logger logger = LogManager.getLogger(ExtensionsRunner.class);
-    public static final TransportInterceptor NOOP_TRANSPORT_INTERCEPTOR = new TransportInterceptor() {
+    private final Logger logger = LogManager.getLogger(ExtensionsRunner.class);
+    private final TransportInterceptor NOOP_TRANSPORT_INTERCEPTOR = new TransportInterceptor() {
     };
 
-    public ExtensionsRunner() throws IOException {}
-
-    public static ExtensionSettings getExtensionSettings() throws IOException {
+    private ExtensionSettings getExtensionSettings() throws IOException {
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java]: That sounds good and the right way. Updated the method name to `readExtensionSettings`

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:61]: Getter and setter are ok (but unnecessary with all-private access) but the original question stands.  It is initialized here to `null` and not set otherwise until the call to `handlePluginsRequest()`.

In `handleIndicesModuleRequest()` you retrieve this value, and in theory it could be null, where it would then be passed to `transportService.connectToNode()` (and later,`transportService.sendRequest()`).  That will probably throw an NPE [here](https://github.com/opensearch-project/OpenSearch/blob/7c73ed8d4c65eee6433e9895036bca1eae4c16e2/server/src/main/java/org/opensearch/transport/TransportService.java#L431).

So should there be a check whether that is null/uninitialized, and a more helpful error message to the user?

Also are there tests that show the preferred order in which these steps are taken?

**Review by saratvemulapalli — COMMENTED** [review]
I did one pass, tiny comments so far.

**Review by saratvemulapalli — COMMENTED** [review]
WIth adding API support for createComponent, can you add unit tests ?

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:61]: That makes sense. So `handlePluginsRequest` will the first request to the SDK from OpenSearch for the transport communication. This request will basically initialize the extension and will be a part of OpenSearch bootstrap. Rest of the request will require a trigger through APIs on OpenSearch side. Tests will covered as a part of https://github.com/opensearch-project/opensearch-sdk/issues/25

File: src/main/java/org/opensearch/sdk/ExtensionClusterStateResponseHandler.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.cluster.ExtensionClusterStateResponse;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponseHandler;
+
+import java.io.IOException;
+
+public class ExtensionClusterStateResponseHandler implements TransportResponseHandler<ExtensionClusterStateResponse> {
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionClusterStateResponseHandler.java:13]: Interesting! This renaming introduced me to an already present `ClusterStateResponse` in [OpenSearch](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/state/ClusterStateResponse.java)

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 18]
```diff
@@ -56,52 +57,88 @@
...
 
+    private void setOpensearchNode(DiscoveryNode opensearchNode) {
+        this.opensearchNode = opensearchNode;
+    }
+
+    private DiscoveryNode getOpensearchNode() {
+        return opensearchNode;
+    }
+
     PluginResponse handlePluginsRequest(PluginRequest pluginRequest) {
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:89]: Sounds right. Created an issue for the same: https://github.com/opensearch-project/opensearch-sdk/issues/28

**Comment by owaiskazi19** [comment]
> WIth adding API support for createComponent, can you add unit tests ?

Abstracted into APIs. Will add tests together as a part of https://github.com/opensearch-project/opensearch-sdk/issues/25

**Review by saratvemulapalli — DISMISSED** [review]
Thanks for taking care of the comments. 
Looks like CI is failing, like due to change not merged to feature/extensions branch on OpenSearch?

**Comment by owaiskazi19** [comment]
> Thanks for taking care of the comments. Looks like CI is failing, like due to change not merged to feature/extensions branch on OpenSearch?

Yes. Build will fail as this PR has dependency on https://github.com/opensearch-project/OpenSearch/pull/3265.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 19]
```diff
@@ -56,52 +56,56 @@
...
     }
 
-    IndicesModuleResponse handleIndicesModuleRequest(IndicesModuleRequest indicesModuleRequest) {
-        logger.info("Indices Module Request");
+    IndicesModuleResponse handleIndicesModuleRequest(IndicesModuleRequest indicesModuleRequest, TransportService transportService) {
+        logger.info("Registering Indices Module Request received from OpenSearch");
         IndicesModuleResponse indicesModuleResponse = new IndicesModuleResponse(true, true, true);
+
+        // CreateComponent
+        DiscoveryNode opensearchNode = getOpensearchNode();
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ExtensionsRunner.java:101]: I still think you need a log warning here if this has not been set. Last comment... will mark this as a suggestion but approve.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:60]: Got it.  I'm going to approve this PR but I marked the spot where I would want a log warning. Or some comment indicating it's needed. Javadocs. Something.  Consider that an "approved with suggestion" as I know you'll be doing more work on this later.

**Review by dbwiddis — APPROVED** [review]
Approved but I really want some placeholder (comment, TODO, log warning, something) about making sure the `opensearchNode` isn't null before using it.  Go ahead and add as an addition for a later issue if needed, just don't want it forgotten.

**Review by dbwiddis — APPROVED** [review]

**Review by joshpalis — APPROVED** [review]
